### DISCUSSION TITLE
Is "docker build ." expected to succeed for this Dockerfile? (Different Docker versions give different results)

### DISCUSSION THREAD

**Discussion by eriksjolund** [moby/moby#44104] [Q&A] [answered]
For this Dockerfile
```
FROM ubuntu:jammy
VOLUME ["/testvol"]
RUN mkdir /testvol
```

I get different results for different Docker versions.

### "docker build ." with Docker desktop on a MacBook Pro succeeds

I installed Docker desktop on a MacBook Pro. With this docker `docker build .` succeeds.

```
$ cat Dockerfile
FROM ubuntu:jammy
VOLUME ["/testvol"]
RUN mkdir /testvol
$ docker build .
[+] Building 1.5s (6/6) FINISHED                                                                     
 => [internal] load build definition from Dockerfile                                            0.0s
 => => transferring dockerfile: 97B                                                             0.0s
 => [internal] load .dockerignore                                                               0.0s
 => => transferring context: 2B                                                                 0.0s
 => [internal] load metadata for docker.io/library/ubuntu:jammy                                 1.5s
 => [1/2] FROM docker.io/library/ubuntu:jammy@sha256:20fa2d7bb4de7723f542be5923b06c4d704370f03  0.0s
 => CACHED [2/2] RUN mkdir /testvol                                                             0.0s
 => exporting to image                                                                          0.0s
 => => exporting layers                                                                         0.0s
 => => writing image sha256:7dc7656f6d4dfe926da7cfb8c5ebf5f3d7574849f575c38d69d13341fc2f92ff    0.0s
$ 
```

```
$ docker --version         
Docker version 20.10.17, build 100c701
```

```
$ docker version  
Client:
 Cloud integration: v1.0.29
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.17.11
 Git commit:        100c701
 Built:             Mon Jun  6 23:04:45 2022
 OS/Arch:           darwin/arm64
 Context:           default
 Experimental:      true

Server: Docker Desktop 4.12.0 (85629)
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.17.11
  Git commit:       a89b842
  Built:            Mon Jun  6 23:01:01 2022
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.8
  GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### "docker build ." on a Fedora CoreOS (rawhide) VM fails

`docker build  .` fails on Fedora CoreOS (rawhide):
```
$ cat Dockerfile 
FROM ubuntu:jammy
VOLUME ["/testvol"]
RUN mkdir /testvol
$ docker build .
Sending build context to Docker daemon  2.048kB
Step 1/3 : FROM ubuntu:jammy
 ---> 21735dab04ba
Step 2/3 : VOLUME ["/testvol"]
 ---> Running in 825c1ffe90b6
Removing intermediate container 825c1ffe90b6
 ---> 58cfd219c155
Step 3/3 : RUN mkdir /testvol
 ---> Running in a0cf517430f7
mkdir: cannot create directory '/testvol': File exists
The command '/bin/sh -c mkdir /testvol' returned a non-zero code: 1
$ 
```

```
$ cat /etc/fedora-release 
Fedora release 38 (Rawhide)
$ docker --version
Docker version 20.10.17, build aa7e414
$ docker version
Client:
 Version:           20.10.17
 API version:       1.41
 Go version:        go1.19
 Git commit:        aa7e414
 Built:             Fri Sep  2 13:47:25 2022
 OS/Arch:           linux/arm64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.17
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.19
  Git commit:       f756502
  Built:            Fri Sep  2 13:47:25 2022
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.6.8
  GitCommit:        
 runc:
  Version:          1.1.3
  GitCommit:        
 docker-init:
  Version:          0.19.0
  GitCommit:        
```

### Another Dockerfile that gives different results

This Dockerfile
```
FROM ubuntu:jammy
VOLUME ["/testvol"]
RUN mkdir -p /testvol/dir
```
builds successfully with the command
```
docker build -t localhost/test .
```
 on both Docker versions  but when I run the container I get different results.

Docker Desktop on Macbook Pro:
```
$ docker run --rm localhost/test ls -ld /testvol/dir
drwxr-xr-x 2 root root 4096 Sep  6 20:36 /testvol/dir
```

Docker in a Fedora CoreOS (rawhide):
```
$ docker run --rm localhost/test ls -ld /testvol/dir
ls: cannot access '/testvol/dir': No such file or directory
```

**Comment by eriksjolund** [comment]
I found two issues that are probably related to this

* https://github.com/moby/buildkit/issues/880
* https://github.com/moby/moby/issues/3639

**Comment by thaJeztah** [comment, marked as answer]
The difference here looks to be that your Docker Desktop installation is using `docker build` with BuildKit enabled, but the Linux installation uses the classic (legacy) builder.

There's a difference in how `VOLUME` instructions are handled between BuildKit and the classic builder.

The classic builder applies volumes defined by `VOLUME` _during_ the `docker build`, which means that any `RUN` that's executed after a `VOLUME` has been defined (which may be in the base image you're using) will have a temporary volume attached.

A result of that was that every content you copy to the `VOLUME` location may not end up in the image, or even in the next `RUN` step.

BuildKit fixed this, by considering the `VOLUME` Dockerfile instruction to be a "configuration" / "metadata-only" step; using `VOLUME` would result in the final image to have a volume definition, but the volume is not used _during_ the build.

So in the classic builder;

```dockerfile
FROM ubuntu:jammy

# this step defines a new volume location in the image's configuration
VOLUME ["/testvol"]

# this step would:
# 1. create a temporary volume
# 2. create a `/testvol` directory inside the image to act as mount-point
# 3. mount the temporary volume at `/testvol`
# 4. execute `mkdir -p /testvol/dir` 
RUN mkdir -p /testvol/dir
```

But note that `/testvol/dir` may not end up in the final image, so someone running the image may not have that directory.

With BuildKit enabled, steps `1. - 3.` do not take place, so the `/testvol` directory wont be there, unless it's explicitly created.


To get the same behaviour both on Docker Desktop and on your Linux machine, I recommend enabling BuildKit; https://docs.docker.com/go/buildkit/

The easiest way to do so is to set the `DOCKER_BUILDKIT=1` environment variable in the shell where the `docker build` is executed (export DOCKER_BUILDKIT=1`, or to prefix the command with `DOCKER_BUILDKIT=1` (`DOCKER_BUILDKIT=1 docker build ........`)

**Reply by eriksjolund** [reply]
Thanks for the explanation. I made a quick fix to the documentation:

* https://github.com/moby/buildkit/pull/3095

(My PR could certainly be improved. I just tried to document the most important difference)

**Reply by thaJeztah** [reply]
Thanks, that's really appreciated! When writing my comment I was indeed wondering if we documented this, so having a PR (even if it's not perfect yet) at least helps as a reminder! 👍
